aiidalab/app.py

```diff
@@ -674,7 +676,7 @@ def stop(self) -> None:
         if self._monitor_thread is not None:
             self._monitor_thread_stop.set()
 
-    def is_alive(self) -> bool | None:
+    def is_alive(self) -> bool | None | Thread:
```

Comment:
I don't think it will return a `Thread` or `None`.

Reply:
`mypy` says otherwise. :sweat_smile: 

```
aiidalab/app.py:681: error: Incompatible return value type
 (got "Thread | bool | None", expected "bool | None")  [return-value]
```

I was extremely surprised by this, but it turns out that in Python `and` and `or` operators do not return bool values, they return one of the operands. See the [reference documentation](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not) and https://stackoverflow.com/a/28321263/3682277.

Reply:
😅. In this case, it will never return `Thread`. But we can keep it if `mypy` complains.

Reply:
You're right, that is weird. I think this is an instance of this issue: https://github.com/python/mypy/issues/16565

In principle, the Thread instance should always be truthy and therefore never returned, *unless* we overwrote its `__bool__` method. I think that's why mypy can't correctly infer this in this case. More info in official docs: https://docs.python.org/3/library/stdtypes.html#truth-value-testing

> By default, an object is considered true unless its class defines either a [__bool__()](https://docs.python.org/3/reference/datamodel.html#object.__bool__) method that returns False or a [__len__()](https://docs.python.org/3/reference/datamodel.html#object.__len__) method that returns zero, when called with the object. [[1]](https://docs.python.org/3/library/stdtypes.html#id12) Here are most of the built-in objects considered false:

Here's a minimal reproducer on mypy playground.
https://mypy-play.net/?mypy=latest&python=3.12&gist=ec46f8172f3e4b8ea65def242d9bc3c4